db/src/db.rs

```diff
@@ -416,45 +437,438 @@ impl DB {
+        let mut non_fts_one: Vec<ReducedEntity> = vec![];
+
+        /// Assertions that are :db.cardinality/many and not :db.fulltext.
+        let mut non_fts_many: Vec<ReducedEntity> = vec![];
+
+        // Transact [:db/add :db/txInstant NOW :db/tx].
+        // TODO: allow this to be present in the transaction data.
+        let now = time::get_time();
+        let tx_instant = (now.sec as i64 * 1_000) + (now.nsec as i64 / (1_000_000));
+        non_fts_one.push((tx,
+                          entids::DB_TX_INSTANT,
+                          TypedValue::Long(tx_instant),
+                          true));
+
+        // Right now, this could be a for loop, saving some mapping, collecting, and type
+        // annotations.  However, I expect it to be a multi-stage map as we start to transform the
+        // underlying entities, in which case this expression is more natural than for loops.
         let r: Vec<Result<()>> = entities.into_iter().map(|entity: &Entity| -> Result<()> {
             match *entity {
                 Entity::Add {
-                    e: entmod::EntidOrLookupRef::Entid(entmod::Entid::Ident(ref e_)),
-                    a: entmod::Entid::Ident(ref a_),
+                    e: entmod::EntidOrLookupRef::Entid(ref e_),
+                    a: ref a_,
                     v: entmod::ValueOrLookupRef::Value(ref v_),
                     tx: _ } => {
 
-                    // TODO: prepare and cache all these statements outside the transaction loop.
-                    // XXX: Error types.
-                    let mut stmt: rusqlite::Statement = conn.prepare("INSERT INTO datoms(e, a, v, tx, value_type_tag, index_avet, index_vaet, index_fulltext, unique_value) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)")?;
-                    let e: i64 = *self.schema.require_entid(&e_.to_string())?;
-                    let a: i64 = *self.schema.require_entid(&a_.to_string())?;
+                    let e: i64 = match e_ {
+                        &entmod::Entid::Entid(ref e__) => *e__,
+                        &entmod::Entid::Ident(ref e__) => *self.schema.require_entid(&e__.to_string())?,
+                    };
+
+                    let a: i64 = match a_ {
+                        &entmod::Entid::Entid(ref a__) => *a__,
+                        &entmod::Entid::Ident(ref a__) => *self.schema.require_entid(&a__.to_string())?,
+                    };
+
+                    let attribute: &Attribute = self.schema.require_attribute_for_entid(&a)?;
+                    if attribute.fulltext {
+                        bail!(ErrorKind::NotYetImplemented(format!("Transacting :db/fulltext entities is not yet implemented: {:?}", entity)))
+                    }
+
+                    // This is our chance to do schema-aware typechecking: to either assert that the
+                    // given value is in the attribute's value set, or (in limited cases) to coerce
+                    // the value into the attribute's value set.
```

Comment:
Add the word TODO? Or do you mean that `to_typed_value` does this to your satisfaction? Clarify the comment, please.